[testing] Refactor to use hypothesis style testing

bwasti opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #19 WIP
* #18 [models] WIP
* #17 [compiler] Add None type handling with sentinel value
* #15 [compiler] Free vars may be less than input vars (some inputs ignored)
* #14 [compiler] Support operators with multiple outputs
* **#16 [testing] Refactor to use hypothesis style testing**

bwasti committed:
[testing] Refactor to use hypothesis style testing

wanchaol commented on test/test_operators.py:
```diff
@@ -1,74 +1,137 @@
+            return F.conv2d(a + a, b, c)
+
+        ref_out, tvm_out = self.runBoth(conv, X, W)
+        assert torch.allclose(ref_out, tvm_out, rtol=0.01, atol=0.01)
```
I think it would be better to standardize the rtol and atol instead of calling them each time, probably make a wrapper with `torch.allclose` and make `rtol` and `atol` be defaults.

bwasti replied:
it's not a huge amount to copy and paste over. I like avoiding wrappers in test cases so that anyone new to the project can copy paste and immediately tweak what they need

wanchaol commented on test/util.py:
```diff
@@ -6,129 +6,80 @@
+                rank_ = cls.rand_int(min_rank, max_rank)()
+            return cls.rand_list(cls.rand_int(min_dim, max_dim), rank_)()
+
+        return get
```
shall we also make a rand_tensor? I think it would be beneficial so that we don't need to construct the tensor everyone we wrote the test.

bwasti replied:
follow up diff? :)

wanchaol reviewed (approved):
looks great! have some nits I think it's good to have, but it's good to go 👍

bwasti committed:
Update on "[testing] Refactor to use hypothesis style testing"

[testing] Refactor to use hypothesis style testing

gh-metadata: pytorch tvm 16 gh/bwasti/5/head

bwasti commented on test/test_operators.py:
```diff
@@ -1,74 +1,137 @@
+        # NCHW
+        X = torch.rand(shape)
+        W = torch.rand((num_kernels, shape[1], kernel_size, kernel_size))
+        bias = torch.rand(num_kernels)
```
what do you mean? like put torch.rand inside the given()?  certainly a good idea :)  perhaps I can do that in a follow up diff
